// @miden-sdk/miden-wallet-adapter-react/dist/useWallet.js
import { createContext, useContext } from 'react';
const EMPTY_ARRAY = [];
const DEFAULT_CONTEXT = {
    autoConnect: false,
    connecting: false,
    connected: false,
    disconnecting: false,
    select(_name) {
        console.error(constructMissingProviderErrorMessage('get', 'select'));
    },
    connect(_privateDataPermission, _network, _allowedPrivateData) {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'connect')));
    },
    disconnect() {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'disconnect')));
    },
    requestTransaction(_transaction) {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'requestTransaction')));
    },
    requestAssets() {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'requestAssets')));
    },
    requestPrivateNotes(_noteFilterType, _noteIds) {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'requestPrivateNotes')));
    },
    signBytes(_message, _kind) {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'signBytes')));
    },
    importPrivateNote(_note) {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'importPrivateNote')));
    },
    requestConsumableNotes() {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'requestConsumableNotes')));
    },
    waitForTransaction(_txId, _timeout) {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'waitForTransaction')));
    },
    requestSend(_transaction) {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'requestSend')));
    },
    requestConsume(_transaction) {
        return Promise.reject(console.error(constructMissingProviderErrorMessage('get', 'requestConsume')));
    },
};
Object.defineProperty(DEFAULT_CONTEXT, 'wallets', {
    get() {
        console.error(constructMissingProviderErrorMessage('read', 'wallets'));
        return EMPTY_ARRAY;
    },
});
Object.defineProperty(DEFAULT_CONTEXT, 'wallet', {
    get() {
        console.error(constructMissingProviderErrorMessage('read', 'wallet'));
        return null;
    },
});
Object.defineProperty(DEFAULT_CONTEXT, 'address', {
    get() {
        console.error(constructMissingProviderErrorMessage('read', 'address'));
        return null;
    },
});
Object.defineProperty(DEFAULT_CONTEXT, 'publicKey', {
    get() {
        console.error(constructMissingProviderErrorMessage('read', 'publicKey'));
        return null;
    },
});
function constructMissingProviderErrorMessage(action, valueName) {
    return ('You have tried to ' +
        ` ${action} "${valueName}"` +
        ' on a WalletContext without providing one.' +
        ' Make sure to render a WalletProvider' +
        ' as an ancestor of the component that uses ' +
        'WalletContext');
}
export const WalletContext = createContext(DEFAULT_CONTEXT);
export function useWallet() {
    return useContext(WalletContext);
}
