// @miden-sdk/miden-wallet-adapter-base/dist/transaction.js
import { u8ToB64 } from './helpers';
export class SendTransaction {
    senderAddress;
    recipientAddress;
    faucetId;
    noteType;
    amount;
    recallBlocks;
    constructor(sender, recipient, faucetId, noteType, amount, recallBlocks) {
        this.senderAddress = sender;
        this.recipientAddress = recipient;
        this.faucetId = faucetId;
        this.noteType = noteType;
        this.amount = amount;
        this.recallBlocks = recallBlocks;
    }
}
export class ConsumeTransaction {
    faucetId;
    noteId;
    noteType;
    amount;
    noteBytes;
    constructor(faucetId, noteId, noteType, amount, noteBytes) {
        this.faucetId = faucetId;
        this.noteId = noteId;
        this.noteType = noteType;
        this.amount = amount;
        this.noteBytes = noteBytes ? u8ToB64(noteBytes) : undefined;
    }
}
export class CustomTransaction {
    address;
    recipientAddress;
    transactionRequest;
    inputNoteIds;
    importNotes;
    constructor(address, recipientAddress, transactionRequest, inputNotesIds, inputNoteBytes) {
        this.address = address;
        this.recipientAddress = recipientAddress;
        const requestBytes = transactionRequest.serialize();
        const base64 = u8ToB64(requestBytes);
        this.transactionRequest = base64;
        this.inputNoteIds = inputNotesIds;
        this.importNotes = inputNoteBytes?.map((note) => u8ToB64(note));
    }
}
export var TransactionType;
(function (TransactionType) {
    TransactionType["Send"] = "send";
    TransactionType["Consume"] = "consume";
    TransactionType["Custom"] = "custom";
})(TransactionType || (TransactionType = {}));
export class Transaction {
    type;
    payload;
    constructor(type, payload) {
        this.type = type;
        this.payload = payload;
    }
    static createSendTransaction(sender, recipient, faucetId, noteType, amount, recallBlocks) {
        const sendTransaction = new SendTransaction(sender, recipient, faucetId, noteType, amount, recallBlocks);
        return new Transaction(TransactionType.Send, sendTransaction);
    }
    static createConsumeTransaction(faucetId, noteId, noteType, amount, noteBytes) {
        const consumeTransaction = new ConsumeTransaction(faucetId, noteId, noteType, amount, noteBytes);
        return new Transaction(TransactionType.Consume, consumeTransaction);
    }
    static createCustomTransaction(address, recipientAddress, transactionRequest, inputNoteIds, noteBytes) {
        const transactionBytes = new CustomTransaction(address, recipientAddress, transactionRequest, inputNoteIds, noteBytes);
        return new Transaction(TransactionType.Custom, transactionBytes);
    }
}
