// @miden-sdk/miden-wallet-adapter-reactui/dist/WalletMultiButton.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useCallback, useEffect, useMemo, useRef, useState, } from 'react';
import { useWallet } from '@miden-sdk/miden-wallet-adapter-react';
import { Button } from './Button';
import { useWalletModal } from './useWalletModal';
import { WalletConnectButton } from './WalletConnectButton';
import { WalletIcon } from './WalletIcon';
import { WalletModalButton } from './WalletModalButton';
export const WalletMultiButton = ({ children, ...props }) => {
    const { address, wallet, disconnect } = useWallet();
    const { setVisible } = useWalletModal();
    const [copied, setCopied] = useState(false);
    const [active, setActive] = useState(false);
    const ref = useRef(null);
    const content = useMemo(() => {
        if (children)
            return children;
        if (!wallet || !address)
            return null;
        const underscoreIndex = address.indexOf('_');
        const frontPart = address.slice(0, 6);
        const middlePart = address.slice(underscoreIndex - 4, underscoreIndex);
        return `${frontPart}...${middlePart}`;
    }, [children, wallet, address]);
    const copyAddress = useCallback(async () => {
        if (address) {
            await navigator.clipboard.writeText(address);
            setCopied(true);
            setTimeout(() => setCopied(false), 400);
        }
    }, [address]);
    const openDropdown = useCallback(() => {
        setActive(true);
    }, []);
    const closeDropdown = useCallback(() => {
        setActive(false);
    }, []);
    const openModal = useCallback(() => {
        setVisible(true);
        closeDropdown();
    }, [setVisible, closeDropdown]);
    useEffect(() => {
        const listener = (event) => {
            const node = ref.current;
            // Do nothing if clicking dropdown or its descendants
            if (!node || node.contains(event.target))
                return;
            closeDropdown();
        };
        document.addEventListener('mousedown', listener);
        document.addEventListener('touchstart', listener);
        return () => {
            document.removeEventListener('mousedown', listener);
            document.removeEventListener('touchstart', listener);
        };
    }, [ref, closeDropdown]);
    if (!wallet)
        return _jsx(WalletModalButton, { ...props, children: children });
    if (!address)
        return _jsx(WalletConnectButton, { ...props, children: children });
    return (_jsxs("div", { className: "wallet-adapter-dropdown", children: [_jsx(Button, { "aria-expanded": active, className: "wallet-adapter-button-trigger", style: {
                    pointerEvents: active ? 'none' : 'auto',
                    width: '100%',
                    backgroundColor: '#FFFFFF',
                    border: '1px solid #D7D7D7',
                    color: 'black',
                    ...props.style
                }, onClick: openDropdown, startIcon: _jsx(WalletIcon, { wallet: wallet }), ...props, children: content }), _jsxs("ul", { "aria-label": "dropdown-list", className: `wallet-adapter-dropdown-list ${active && 'wallet-adapter-dropdown-list-active'}`, ref: ref, role: "menu", children: [_jsx("li", { onClick: copyAddress, className: "wallet-adapter-dropdown-list-item", role: "menuitem", children: copied ? 'Copied' : 'Copy address' }), _jsx("li", { onClick: openModal, className: "wallet-adapter-dropdown-list-item", role: "menuitem", children: "Change wallet" }), _jsx("li", { onClick: disconnect, className: "wallet-adapter-dropdown-list-item", role: "menuitem", children: "Disconnect" })] })] }));
};
